dns-integrator-operator/src/charm.py

```diff
@@ -0,0 +1,98 @@
+
+logger = logging.getLogger(__name__)
+
+VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+
+
+class DnsIntegratorOperatorCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: typing.Any):
+        """Construct.
+
+        Args:
+            args: Arguments passed to the CharmBase parent constructor.
+        """
+        super().__init__(*args)
+        self.dns_record = dns_record.DNSRecordRequires(self)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle changes in configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring charm")
+        logger.debug("UPDATE INTEGRATOR CONFIG")
+        self._update_relations()
+        self.unit.status = ops.ActiveStatus()
+
+    def _update_relations(self) -> None:
+        """Update all DNS data for the existing relations."""
+        if not self.model.unit.is_leader():
+            return
+        try:
+            for relation in self.model.relations[self.dns_record.relation_name]:
+                self.dns_record.update_relation_data(
+                    relation, self._get_dns_record_requirer_data()
+                )
+        except ops.model.ModelError as e:
+            logger.error("ERROR while updating relation data: %s", e)
+
+    def _get_dns_record_requirer_data(self) -> dns_record.DNSRecordRequirerData:
+        """Get DNS record requirer data."""
+        entries = []
+        for request in str(self.config["requests"]).split("\n"):
+            data = request.split()
+            if len(data) != 6:
+                continue
+            (host_label, domain, ttl, record_class, record_type, record_data) = data
+            entry = dns_record.RequirerEntry(
+                host_label=host_label,
+                domain=domain,
+                ttl=ttl,
+                record_class=record_class,
+                record_type=record_type,
+                record_data=record_data,
+                uuid=self._uuidv4(" ".join(data)),
+            )
+            entries.append(entry)
+        return dns_record.DNSRecordRequirerData(dns_entries=entries)
+
+    def _uuidv4(self, seed: str = None) -> str:
+        """Get stable uuid.
+
+        Args:
+            seed: string seed used to create the output
+
+        Returns:
+            UUID constructed based on the given seed
+        """
+        if seed is None:
+            return str(uuid.uuid4())
+
+        hash = hashlib.sha512(seed.encode())
+        hash_num = int.from_bytes(hash.digest(), byteorder='big')
+        hash_str = hash.hexdigest()
+
+        # Determine the variant character (8,9,a,b)
+        variant_chars = ["8", "9", "a", "b"]
+        variant = variant_chars[hash_num % 4]
+
+        # Construct the UUIDv4
+        return f"{hash_str[0:8]}-{hash_str[8:12]}-4{hash_str[12:15]}-{variant + hash_str[15:18]}-{hash_str[18:30]}"
```

Comment:
I'm not familiar with this but why not use uuid python package instead?

Reply:
(this function is really cool BTW)

Reply:
I need a UUID that is constructed based on the record request to always have the same. This was the bug on the last demo. If I don't do that, the uuid of the request will always change and be added as a new one in the DNS policy operator or the bind operator

Reply:
I'm falling back to python's uuid generation if the seed is empty (but that shouldn't really be used in practice)

Reply:
Could you add a comment about that explaining why the uuid package is not being used?

Reply:
Added an explanation (and forced the seed usage)